### PR TITLE
ADD: mlm memory improvements

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/nn/target_blocks.py [hunk 1]
```diff
@@ -64,19 +63,43 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
-        # )
-        # final_proj_b = tf.get_variable(
-        #     'dense/bias',
-        #     [config.n_embed],
-        #     initializer=tf.zeros_initializer
-        # )
-        # final_proj = act_fns[config.act_fn](
-        #     tf.matmul(gathered_hidden, final_proj_w, transpose_b=True) + final_proj_b
-        # )
+def masked_language_model(*, X, M, mlm_weights, mlm_ids, mlm_positions, embed_weights, hidden, config, reuse=None, train=False):
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:49]: you probably want to delete the previous declaration of a function of the same name? (line 50)

File: finetune/nn/target_blocks.py [hunk 2]
```diff
@@ -64,19 +63,43 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
+    
+    with tf.variable_scope('model/masked-language-model'):
+        batch, seq, feats = shape_list(hidden)
+        flat_offsets = tf.reshape(
+            tf.range(0, batch, dtype=tf.int32) * seq, [-1, 1]
+        )
+
+        not_padding = tf.reshape(mlm_weights, [-1]) > 1e-9
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:57]: why >1e-9 here? aren't the weights typically 0 or 1? even just >0 should be fine?

File: finetune/target_models/masked_language_model.py [hunk 3]
```diff
@@ -62,14 +62,18 @@ def text_to_tokens_mask(self, X, Y=None, context=None):
                     ]
                 )
             ] = False
+            mlm_positions = np.where(mlm_mask)[0]
+            
+            if len(mlm_positions) > self.config.max_masked_tokens: # subsample
+                np.random.shuffle(mlm_positions) # means we don't bias the begining of the sequence
+                mlm_positions = mlm_positions[:self.config.max_masked_tokens]
```
Comment (lhz1029) [inline finetune/target_models/masked_language_model.py]: nitpick, but probably more efficient to do something like this?

positions_to_keep = np.random.choice(range(seq_len), size=self.config.max_masked_tokens)
mlm_positions = mlm_positions[positions_to_keep]

File: finetune/target_models/masked_language_model.py [hunk 4]
```diff
@@ -99,6 +103,9 @@ class MaskedLanguageModel(BaseModel):
     :param config: A :py:class:`finetune.config.Settings` object or None (for default config).
     :param \**kwargs: key-value pairs of config items to override.
     """
+
+    defaults = dict(low_memory_mode=True, batch_size=48, xla=True)
```
Comment (lhz1029) [inline finetune/target_models/masked_language_model.py:106]: you've found that these are better than the config.py defaults in terms of speed?

File: finetune/nn/target_blocks.py [hunk 5]
```diff
@@ -64,19 +63,43 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
+        flat_positions = tf.boolean_mask(tf.reshape(mlm_positions + flat_offsets, [-1]), not_padding) # take off the padding entirely
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:58]: why do you need the boolean mask? maybe I'm missing something, but doesn't `tf.reshape(mlm_positions + flat_offsets, [-1])` just give the positions, offset by the example in the batch?

Comment (lhz1029) [reply finetune/nn/target_blocks.py:58]: also, it looks like mlm_weights is just a tensor of 1s now, so the mask probably doesn't do anything?

File: finetune/nn/target_blocks.py [hunk 6]
```diff
@@ -88,17 +111,16 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
-        mlm_weights = tf.reshape(mlm_weights, [-1])
-
-        log_probs = tf.nn.log_softmax(logits, axis=-1)
-        one_hot_labels = tf.one_hot(mlm_ids, depth=n_vocab, dtype=tf.float32)
-        per_example_loss = -tf.reduce_sum(log_probs * one_hot_labels, axis=[-1])
-        numerator = tf.reduce_sum(mlm_weights * per_example_loss)
-        denominator = tf.reduce_sum(mlm_weights) + 1e-5
-        mlm_loss = numerator / denominator
-
+        mlm_loss = tf.contrib.losses.sparse_softmax_cross_entropy(
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:97]: nice!

File: finetune/nn/target_blocks.py [hunk 7]
```diff
@@ -88,17 +111,16 @@ def masked_language_model(*, X, M, mlm_weights, mlm_positions, mlm_ids, embed_we
...
-        numerator = tf.reduce_sum(mlm_weights * per_example_loss)
-        denominator = tf.reduce_sum(mlm_weights) + 1e-5
-        mlm_loss = numerator / denominator
-
+        mlm_loss = tf.contrib.losses.sparse_softmax_cross_entropy(            
+            logits,
+            mlm_ids,
+        ) # No weights needed as there is no padding.
+
+        logits = tf.scatter_nd(
```
Comment (lhz1029) [inline finetune/nn/target_blocks.py:102]: :+1:

**Review by lhz1029 — APPROVED** [review]

File: finetune/nn/target_blocks.py (hunk 5 above)
Comment (benleetownsend) [reply finetune/nn/target_blocks.py:58]: So, any padding added by padded_batch. this removes it.

**Comment by madisonmay** [comment]
LGTM sans deleting the old masked_language_model fn

**Comment by benleetownsend** [comment]
Notes:
xla seems to fail with the concatenated variables.
Reduced batch size needed when context is included due to memory footprint increases

**Comment by benleetownsend** [comment]
Superseded by other PRs
